src/aws/flb_aws_util.c

```diff
@@ -507,3 +511,171 @@ int flb_read_file(const char *path, char **out_buf, size_t *out_size)
 
     return 0;
 }
+
+/* Generic replace function for strings. */
+char* replace_uri_tokens(const char* original_string, const char* current_word, 
+                         const char* new_word) 
+{ 
+    char *result; 
+    int i = 0;
+    int count = 0; 
+    int new_word_len = strlen(new_word); 
+    int old_word_len = strlen(current_word); 
+  
+    for (i = 0; original_string[i] != '\0'; i++) { 
+        if (strstr(&original_string[i], current_word) == &original_string[i]) { 
+            count++; 
+            i += old_word_len - 1; 
+        } 
+    }
+
+    result = flb_sds_create_size(i + count * (new_word_len - old_word_len) + 1);
+    if (!result) {
+        flb_errno();
+        return NULL;
+    }
+
+    i = 0; 
+    while (*original_string) { 
+        if (strstr(original_string, current_word) == original_string) { 
+            strcpy(&result[i], new_word); 
+            i += new_word_len; 
+            original_string += old_word_len; 
+        } 
+        else
+            result[i++] = *original_string++; 
+    } 
+
+    result[i] = '\0'; 
+    return result; 
+} 
+
+/* Constructs S3 object key as per the format. */
+flb_sds_t get_s3_key(const char *format, time_t time, char *tag, char *tag_delimiter)
+{
+    int i = 0;
+    int ret = 0;
+    char *tag_token = NULL;
+    flb_sds_t tmp = NULL;
+    flb_sds_t buf = NULL;
+    flb_sds_t s3_key = NULL;
+    flb_sds_t tmp_key = NULL;
+    flb_sds_t tmp_tag = NULL;
+    struct tm *gmt = NULL;
+    
+    if (strlen(format) > S3_KEY_SIZE){
+        flb_warn("[s3_key] Object key length is longer than the 1024 character limit.");
+        return NULL;
+    }
+
+    tmp_tag = flb_sds_create_len(tag, strlen(tag) + 1);
+    if(!tmp_tag){
+        flb_errno();
+        return NULL;
+    }
+
+    s3_key = flb_sds_create_len(format, S3_KEY_SIZE);
+    if (!s3_key) {
+        flb_errno();
+        return NULL;
+    }
+
+    /* Check if delimiter(s) specifed exists in the tag. */
+    for (i = 0; i < strlen(tag_delimiter); i++){
+        if (strchr(tag, tag_delimiter[i])){
+            ret = 1;
+            break;
+        } 
+    }
+
+    if(ret == 0){
+        flb_warn("[s3_key] Invalid Tag delimiter: does not exist in tag.");
+        flb_sds_destroy(tmp_tag);
+        flb_sds_destroy(s3_key);
+        return NULL;
+    }  
+
+    /* Split the string on the delimiters */
+    tag_token = strtok(tmp_tag, tag_delimiter);
+    
+    /* Find all occurences of $TAG[*] and  
+     * replaces it with the right token from tag.
+     */
```

Comment:
nit, the first call to strtok is part of the logic of the loop. Move it below this comment to right before the loop. It doesn't really need its own comment either, its fairly obvious from the function parameters, and anyone could look up the function if they don't know it.

Reply:
Actually wait I think I am being dumb- the first call to strtok is different than the others

Reply:
since the loop using null

Reply:
Yeah disregard this comment

Reply:
I made this change to resolve https://github.com/fluent/fluent-bit/pull/2467#discussion_r471968636

Reply:
Sorry